qcodes/dataset/data_set.py

```diff
@@ -457,6 +492,10 @@ def add_result(self, results: Dict[str, VALUE]) -> int:
         the name of a parameter in this DataSet.
         It is an error to add results to a completed DataSet.
         """
+
+        if not self.started:
```

Comment:
this is going to slow things down because `self.number_of_results` is a query :( how about caching the "started" fact?

Reply:
Before checking number of results:
```
=========================================================== 
                                         param1                                     
              ----------------------------------------------------------- ----------
                {'n_values': 10000, 'n_times': 2, 'paramtype': 'array'}    14.9±0ms 
                {'n_values': 100, 'n_times': 200, 'paramtype': 'array'}    219±0ms  
               {'n_values': 10000, 'n_times': 2, 'paramtype': 'numeric'}   631±0ms  
               {'n_values': 100, 'n_times': 200, 'paramtype': 'numeric'}   648±0ms  
              =========================================================== 
```
Now:
```
 =========================================================== 
                                         param1                                     
              ----------------------------------------------------------- ----------
                {'n_values': 10000, 'n_times': 2, 'paramtype': 'array'}    19.2±0ms 
                {'n_values': 100, 'n_times': 200, 'paramtype': 'array'}    227±0ms  
               {'n_values': 10000, 'n_times': 2, 'paramtype': 'numeric'}   688±0ms  
               {'n_values': 100, 'n_times': 200, 'paramtype': 'numeric'}   695±0ms  
              =========================================================== 
```
but I don't know what the natural variation is. I'll look into caching.